File: ios/OakApp/Networking/SSEParser.swift
import Foundation

/// A pure, incremental Server-Sent-Events line parser: it is fed `event:`/`data:`
/// field lines one at a time and emits typed events on each blank-line-terminated
/// frame. The byte→line splitting (``ByteLineSplitter``) and the stream loop
/// (``SSEClient``) are shared across event families; this seam is what lets a
/// SIBLING parser (e.g. the team-builder's ``BuilderSSEParser``, whose terminal
/// `answer` carries a `BuilderAnswer` not an `OakAnswer`) reuse all of that plumbing
/// while decoding its own event union.
///
/// `Sendable` so the shared stream loop can create one inside its `@Sendable` task.
protocol SSELineParser: Sendable {
  /// The event type this parser decodes each frame into.
  associatedtype Event: Sendable

  /// Feeds one line. Returns the events completed by this line (zero or one for a
  /// well-formed stream — a blank line that closes a frame).
  mutating func consume(line: String) throws -> [Event]

  /// Flushes any buffered frame at end-of-stream (a defensive backstop; well-formed
  /// streams end with a blank line, so the buffer is already empty).
  mutating func finish() throws -> [Event]
}

/// A pure, incremental Server-Sent-Events frame parser (component-design.md
/// "Networking layer"). It accumulates `event:`/`data:` field lines, emits one
/// ``SSEEvent`` per blank-line-terminated frame, and ignores `:`-prefixed comment
/// lines (the `: keep-alive` heartbeats sent every 15s).
///
/// **No I/O** — the caller feeds it lines (e.g. from `URLSession.AsyncBytes.lines`,
/// which already strips newlines) one at a time; the parser owns only the in-flight
/// frame state. This keeps it trivially unit-testable against recorded `.sse`
/// fixtures.
///
/// Decoding policy:
///   * A recognized event whose `data:` JSON fails to decode throws
///     `OakError.decoding(...)` (a contract drift the consumer surfaces) — this is
///     the only thing that makes the stream fail from inside the parser.
///   * An **unknown** event name is ignored (forward-compatible with new events).
///   * Comment lines, unknown fields (`id`/`retry`), and empty/incomplete frames
///     (e.g. a trailing blank line) emit nothing.
struct SSEParser: SSELineParser {
  private var eventName: String?
  private var dataBuffer = ""
  private var hasData = false

  init() {}

  /// Feeds one line. Returns the events completed by this line (zero or one for a
  /// well-formed stream — a blank line that closes a frame).
  mutating func consume(line rawLine: String) throws -> [SSEEvent] {
    // Robustness against CRLF input even though `.lines` normalizes newlines.
    var line = rawLine
    if line.hasSuffix("\r") {
      line.removeLast()
    }

    if line.isEmpty {
      return try dispatch()
    }
    if line.hasPrefix(":") {
      return []  // comment / heartbeat
    }

    let (field, value) = Self.splitField(line)
    switch field {
    case "event":
      eventName = value
    case "data":
      if hasData {
        dataBuffer.append("\n")
      }
      dataBuffer.append(value)
      hasData = true
    default:
      break  // id / retry / unknown field → ignored per SSE spec
    }
    return []
  }

  /// Flushes any buffered frame at end-of-stream. Well-formed streams end with a
  /// blank line so the buffer is already empty; this is a defensive backstop.
  mutating func finish() throws -> [SSEEvent] {
    try dispatch()
  }

  // MARK: - Internals

  /// Closes the in-flight frame, decodes it into an ``SSEEvent``, and resets state.
  private mutating func dispatch() throws -> [SSEEvent] {
    let name = eventName
    let payload = dataBuffer
    let complete = hasData
    reset()

    guard let name, complete else {
      return []  // empty/incomplete frame (e.g. a trailing blank line)
    }

    let json = Data(payload.utf8)
    let decoder = JSONDecoder()
    do {
      switch name {
      case "turn":
        let data = try decoder.decode(SSEEvent.TurnData.self, from: json)
        return [.turn(turnId: data.turnId)]
      case "scope":
        let data = try decoder.decode(SSEEvent.ScopeData.self, from: json)
        return [.scope(format: data.format, source: data.source)]
      case "tool_activity":
        let data = try decoder.decode(SSEEvent.ToolActivityData.self, from: json)
        return [.toolActivity(tool: data.tool, label: data.label)]
      case "answer_start":
        return [.answerStart]
      case "answer_delta":
        let data = try decoder.decode(SSEEvent.AnswerDeltaData.self, from: json)
        return [.answerDelta(text: data.text)]
      case "answer":
        let data = try decoder.decode(SSEEvent.AnswerData.self, from: json)
        return [.answer(data.answer)]
      case "error":
        let data = try decoder.decode(SSEEvent.ErrorData.self, from: json)
        return [.error(code: data.code, message: data.message, status: data.status)]
      case "stopped":
        return [.stopped]  // terminal; `data: {}` carries no payload
      default:
        return []  // unknown event name → forward-compatible no-op
      }
    } catch {
      throw OakError.decoding("SSEEvent.\(name)")
    }
  }

  /// Resets the in-flight frame state.
  private mutating func reset() {
    eventName = nil
    dataBuffer = ""
    hasData = false
  }

  /// Splits a `field: value` SSE line, stripping a single optional leading space
  /// from the value (per the SSE spec). A line with no colon is a field with an
  /// empty value.
  private static func splitField(_ line: String) -> (field: String, value: String) {
    guard let colon = line.firstIndex(of: ":") else {
      return (line, "")
    }
    let field = String(line[line.startIndex..<colon])
    var valueStart = line.index(after: colon)
    if valueStart < line.endIndex, line[valueStart] == " " {
      valueStart = line.index(after: valueStart)
    }
    return (field, String(line[valueStart..<line.endIndex]))
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/networking/SseParser.kt
package ai.gowtam.oak.networking

import ai.gowtam.oak.wire.Format
import ai.gowtam.oak.wire.OakAnswer
import ai.gowtam.oak.wire.OakJson
import ai.gowtam.oak.wire.ScopeSource
import ai.gowtam.oak.wire.SseEvent
import kotlinx.serialization.SerializationException
import kotlinx.serialization.Serializable

/**
 * The contract implemented by [SseParser] and [BuilderSseParser] (component-design.md
 * "Networking layer", mirrors iOS `SSELineParser`): feed one line at a time and get back
 * the events completed by that line (zero or one for a well-formed stream — a blank line
 * that closes a frame); [finish] flushes any buffered frame at end-of-stream (a
 * defensive backstop — well-formed streams end on a blank line, so the buffer is already
 * empty). [SseClient]'s stream loop is generic over this seam so the SAME byte/frame
 * plumbing serves the chat and team-builder event families with different event unions.
 */
interface SseLineParser<T> {
    /** Feeds one line (already newline-stripped by [ByteLineSplitter]). */
    fun consume(line: String): List<T>

    /** Flushes any buffered frame at end-of-stream. */
    fun finish(): List<T>
}

/**
 * One completed SSE frame's raw materials: the `event:` name and the joined `data:`
 * payload (each `data:` line joined by `\n` per the SSE spec).
 */
internal data class RawSseFrame(val name: String, val payload: String)

/**
 * Frame-accumulation logic shared by [SseParser] and [BuilderSseParser] so the
 * `event:`/`data:` line framing (identical for both event families) isn't duplicated —
 * only the per-event JSON decoding differs between the two. Pure and synchronous: no
 * I/O, no knowledge of any event schema.
 *
 * Accumulates field lines fed one at a time and completes a [RawSseFrame] on the blank
 * line that delimits an SSE event. Comment lines (a leading `:`, e.g. the `: keep-alive`
 * heartbeat sent every 15s) and unknown fields (`id`/`retry`) are ignored per the SSE
 * spec. A trailing `\r` left by [ByteLineSplitter] on CRLF input is stripped here.
 */
internal class SseFrameAccumulator {
    private var eventName: String? = null
    private val dataBuffer = StringBuilder()
    private var hasData = false

    /** Feeds one line. Returns the completed frame, or `null` if the frame isn't closed yet. */
    fun consume(rawLine: String): RawSseFrame? {
        val line = if (rawLine.endsWith("\r")) rawLine.dropLast(1) else rawLine

        if (line.isEmpty()) return dispatch()
        if (line.startsWith(":")) return null // comment / heartbeat

        val (field, value) = splitField(line)
        when (field) {
            "event" -> eventName = value
            "data" -> {
                if (hasData) dataBuffer.append('\n')
                dataBuffer.append(value)
                hasData = true
            }
            else -> Unit // id / retry / unknown field → ignored per SSE spec
        }
        return null
    }

    /** Flushes any buffered frame at end-of-stream. */
    fun finish(): RawSseFrame? = dispatch()

    private fun dispatch(): RawSseFrame? {
        val name = eventName
        val payload = dataBuffer.toString()
        val complete = hasData
        reset()
        if (name == null || !complete) return null // empty/incomplete frame (e.g. a trailing blank line)
        return RawSseFrame(name, payload)
    }

    private fun reset() {
        eventName = null
        dataBuffer.setLength(0)
        hasData = false
    }

    /**
     * Splits a `field: value` SSE line, stripping a single optional leading space from
     * the value (per the SSE spec). A line with no colon is a field with an empty value.
     */
    private fun splitField(line: String): Pair<String, String> {
        val colon = line.indexOf(':')
        if (colon < 0) return line to ""
        var valueStart = colon + 1
        if (valueStart < line.length && line[valueStart] == ' ') valueStart++
        return line.substring(0, colon) to line.substring(valueStart)
    }
}

/**
 * A pure, incremental Server-Sent-Events frame parser for the chat stream (mirrors iOS
 * `SSEParser`; component-design.md "Networking layer"). It delegates the `event:`/`data:`
 * line framing to [SseFrameAccumulator] and decodes each completed frame's `data:` JSON
 * into an [SseEvent].
 *
 * **No I/O** — the caller ([SseClient]) feeds it lines fed off [ByteLineSplitter] one at a
 * time; this keeps the parser trivially unit-testable against recorded `.sse` fixtures.
 *
 * Decoding policy:
 *  - a recognized event whose `data:` JSON fails to decode throws [OakError.Decoding] (a
 *    contract drift the consumer surfaces) — the only thing that fails the stream from
 *    inside the parser;
 *  - an **unknown** event name is ignored (forward-compatible with new events);
 *  - comment lines, unknown fields, and empty/incomplete frames emit nothing.
 */
class SseParser : SseLineParser<SseEvent> {
    private val frames = SseFrameAccumulator()

    override fun consume(line: String): List<SseEvent> = dispatch(frames.consume(line))

    override fun finish(): List<SseEvent> = dispatch(frames.finish())

    private fun dispatch(frame: RawSseFrame?): List<SseEvent> {
        val (name, payload) = frame ?: return emptyList()
        return try {
            when (name) {
                "scope" -> {
                    val data = OakJson.decodeFromString(ScopeData.serializer(), payload)
                    listOf(SseEvent.Scope(format = data.format, source = data.source))
                }
                "tool_activity" -> {
                    val data = OakJson.decodeFromString(ToolActivityData.serializer(), payload)
                    listOf(SseEvent.ToolActivity(tool = data.tool, label = data.label))
                }
                "answer_start" -> listOf(SseEvent.AnswerStart)
                "answer_delta" -> {
                    val data = OakJson.decodeFromString(AnswerDeltaData.serializer(), payload)
                    listOf(SseEvent.AnswerDelta(text = data.text))
                }
                "answer" -> {
                    val data = OakJson.decodeFromString(AnswerData.serializer(), payload)
                    listOf(SseEvent.Answer(data.answer))
                }
                "error" -> {
                    val data = OakJson.decodeFromString(ErrorData.serializer(), payload)
                    listOf(SseEvent.Error(code = data.code, message = data.message, status = data.status))
                }
                else -> emptyList() // unknown event name → forward-compatible no-op
            }
        } catch (e: SerializationException) {
            throw OakError.Decoding("SseEvent.$name")
        } catch (e: IllegalArgumentException) {
            throw OakError.Decoding("SseEvent.$name")
        }
    }

    @Serializable
    private data class ScopeData(val format: Format, val source: ScopeSource)

    @Serializable
    private data class ToolActivityData(val tool: String, val label: String)

    @Serializable
    private data class AnswerDeltaData(val text: String)

    @Serializable
    private data class AnswerData(val answer: OakAnswer)

    @Serializable
    private data class ErrorData(val code: String, val message: String, val status: Int? = null)
}
